### ISSUE TITLE
remove unpopular system extensions

### ISSUE THREAD

**Issue by guicho271828** [guicho271828/lisp-interface-library#20]
Make it more simpler!

+ [ ] uiop:define-package
+ [ ] package-inferred-systems

**Comment by fare** [comment]
define-package's :mix is really useful in some cases.
define-package also makes it much nicer to edit and/or upgrade code in presence of package redefinitions, which tends to happen a lot with package-inferred-systems.

package-inferred-systems is a very nice style of writing CL code, are you proposing to move back to the throw-everything-in-a-big-package model?

**Comment by guicho271828** [comment]
I agree that splitting the packages is a good idea in general. However, the problem which makes me keep away from define-package is that macros generally lack the composability. This is why many `defclass*` or `defpackage*` and alike are the bad ideas. Unless it is extensible just like `iterate` has  `defclause-sequence`, I might be more pleased to see its functional versions. For example, 

```
(cl:defpackage XXX .... )
(setf (package XXX) (mix (package YYY) (package ZZZ)))
```

Well, at least my rants in this issues list are for my own use, and I don't plan to do all of this because LIL and p-i-systems are at least working and there is no need to break & change it right now.

**Comment by fare** [comment]
That's an interesting point of view, and in general, I'd approve, but in the case of defpackage at least, there is a big problem: it is an error when redefining a package with defpackage if the new definition doesn't exactly match the previous. That's an issue while developing, and an even bigger isssue when doing a live upgrade of code (which is what define-package was meant to enable, for asdf itself).

Now that you mention it, the correct solution would have been a composable interface *inside* the package definition macro: you'd compose functions (optionally designated by keywords) that, applied to a package specification, return a modified package specification (if possible, in a pure functional way). In the end, the composed functions are applied to an empty specification, and a package is created from the final specification. Extensibility is then achieved by adding new functions to this protocol and/or extending the specification class in an OO way.

(def package foo (:use :cl) (:nicknames nick1 nick2) ...)

Considering the early place of define-package in the asdf bootstrap process, though, such a design would not be appropriate for it. But certainly, there is an opportunity for a better package definition protocol, and once it exists, it might be appropriate for lil to use it.
